Big O (#39)
Ada-Developers-Academy/core-unit-1: big-o/big-o-analysis.md

- apply it to a large example, and give explanations

We encourage you all to read through these sections. After each example, re-read the example and follow along the same thought process, so you can come to the same conclusions too.

## Time Complexity: Code to Big O

Recall that time complexity is the measurement of how the amount of time an algorithm takes to run as the size of the input changes. Also, time complexity is usually measured in _number of operations._ Therefore, our goal is to find a formula that captures how many operations an algorithm executes.

### Generic Steps

1. Read through the code, and identify all lists that have a variable size.
   - Typically, with one list, we say it has _n_ number of elements.
1. Identify all of the operations in the algorithm
1. Recognize which operations are related to the lists of size _n_
   - Typically, the operations inside of a for loop get multipled by _n_ times
   - This relationship will change depending on the kind of loop and the logic in the loop. Consider cases where the loop `break`s.
1. Create an equation that represents how many operations there are
   - This equation can use _n_ as a variable

> 1. Drop the constants (notes below)

Author: I'll do a read-over of this entire lesson; if it feels like there's room (aka, I wouldn't predict the student's brains are already overflowing with new info in a not-useful-way), I'll add a paragraph in the Constants section about the highest value.

   1. `i += 1`
   1. `j -= 1`

3. Recognize which operations have a relationship to list size:

   1. All 4 of the operations outside of the `while` loop will only happen once; they have no relationship to list size
   1. All 6 of the operations inside the `while` loop can happen _n_ times

4. Create a formula that counts how many operations there are:

   1. We can express this as _4 + (6 \* n)_, or _4 + 6n_

5. Drop the constants:

   1. We are able to drop `4 +` and `6 *`
   1. After dropping the constants, our remaining formula is _n_

6. Match this formula with the most relevant complexity:
   1. _n_ matches _O(n)_ the most closely. This algorithm has linear time complexity.

### What is "Dropping the Constant," or Finding the Dominant Order?

In complexity analysis, programmers care about the dominant order, or the part of the complexity that affect the complexity _the most._

Mathematical constants are values that have a fixed value. Constants tend to be added, subtracted, multiplied, or divided in our formulas. In Big O, constants will affect our complexity (O(6n) is different than O(n))... but they do not affect the curve enough that is meaningful for complexity analysis.

> 

Reviewer: Constants are emphasized, but this also good place to mention that all terms except the largest are dropped.

Some examples:

- O(2n) is shortened to O(n)
- O(n/2) (or one half of _n_) is shortened to O(n)
- O(n + n<sup>2</sup>) is shortened to O(n<sup>2</sup>)

> 

Reviewer 2: I think we should move this closer to the first example. Students will likely be curious about dropping the constant there.

## Space Complexity: Code to Big O

> Recall that space complexity is the measurement of how much memory an algorithm uses as the size of the input changes. Space complexity is measured in the _amount of memory allocated in this algorithm._ This amount of memory doesn't include the memory that stores our data set. Our goal will be to count the number of variables needed.

Reviewer: > Our goal will be to count the number of variables needed.

It's definitely tricky to find a good, simple assertion about what you're looking for with space analysis for students. New variables can be a sign of memory usage but not always and there are plenty of ways to consume memory without assigning variables.

I've had some luck communicating in terms of creating new things or values. E.g. "We're looking for places where the code creates new values like creating new lists, creating new objects, strings, primitives."

Author: This suggestion is really helpful and awesome! Thank you!!

```suggestion
Recall that space complexity is the measurement of how much memory an algorithm uses as the size of the input changes. Space complexity is measured in the _amount of memory allocated in this algorithm._ This amount of memory doesn't include the memory that stores our data set.

Our goal will be to count the number of new values that are stored. We'll look for places where the code creates new values, like creating new lists, new objects, strings, or primitives.

A great strategy is to count the number of initialized variables, but it will be great to understand that this is only one factor of space complexity.
```

We will get into the details of lists and memory in a future lesson.

### !end-callout

### Generic Steps

1. Read through the code, and identify all places where a variable is _initially assigned._
   - We do not need to count re-assignment of variables, as the value will occupy the same memory as the initial assignment
1. Recognize which variables have a value that could take a variable amount of memory
   - Typically, storing a list will require _n_ amount of memory
1. Create an equation that represents how many values are initialized and stored in memory
1. Drop the constants
1. Match this Big O to a relevant complexity

### Small Example

Given an implementation of linear search...

```python
def linear_search(array, item):
    for i in range(len(array)):
        if item == array[i]:
            return i
    return False
```

1. Identify all places where the algorithm initializes a variable. (Don't count the input data set!)

   - `i`

>    - `item`

Reviewer: if the input data set (`array`, here) isn't counted then `item` probably shouldn't be either.

Author: Good call! I'll include both with a parenthetical, since folks will naturally question its presence if we don't include it.

```suggestion
   - `item`
   - `array`

Note: For this exercise and the purpose of illustrating space complexity, we'll consider `item` and won't consider `array`. However, it is a completely valid interpretation to not consider `item` or `array`, as they are "inputs" into this function.
```

1. Recognize which operations are related to the lists of size _n_
   - Typically, the operations inside of a for loop get multipled by _n_ times
   - This relationship will change depending on the kind of loop and the logic in the loop. Consider cases where the loop `break`s.
1. Create an equation that represents how many operations there are
   - This equation can use _n_ as a variable
1. Drop the constants (notes below)
1. Match this Big O to the most relevant complexity

### !callout-secondary

## What's an Operation?

One operation is when one operator is used. There are arithmetic operators (`+`, `-`, `*`, `%`, etc), assignment operators (`=`, `+=`, etc), comparison operators (`==`, `<`, `>=`, etc), logical operators (`and`, `or`, `not`), and more miscellaneous ones (`is`, `is not`, `in`, `not in`, etc)

### !end-callout

### Small Example

This is one implementation of the linear search algorithm. It takes in a list of _n_ element, and an item to search for. It's responsibility is to return the index of that item.

```python
def linear_search(array, item):
    for i in range(len(array)):
        if item == array[i]:
            return i
    return False
```

> 1. Identify all lists with variable size: There is one list with variable size, and it's a size of _n_.

Author: ```suggestion
1. Identify all lists with variable size: There is one list with variable size, `array`, and it's a size of _n_.
```
